### ISSUE TITLE
SERVFAIL

### ISSUE THREAD

**Issue by mirazsarker** [zmap/zdns#309]
i am faceing this

root@vmi697131:~/zdns#  echo "199.59.243.200" |  ./zdns alookup
{"name":"199.59.243.200","status":"SERVFAIL","timestamp":"2022-06-02T19:47:14+02:00"}
root@vmi697131:~/zdns# echo "censys.io" | ./zdns mxlookup --ipv4-lookup
{"name":"censys.io","status":"SERVFAIL","timestamp":"2022-06-02T19:50:40+02:00"}

**Comment by PenelopeFudd** [comment]
I'm getting SERVFAIL too.

I figured out why:
* If not specified, zdns uses the nameserver listed in the file `/etc/resolv.conf`
* On Ubuntu 18.04 and 20.04, that is 127.0.0.53, the address of the systemd resolver called systemd-resolved.
* When zdns sends a request to 127.0.0.53, it should use the source ip address 127.0.0.1, but instead uses the default address for the machine
* systemd-resolved generates the error and returns the SERVFAIL message to zdns: 
```
Nov 03 12:17:26 ns3.example.com systemd-resolved[508]: Got packet on unexpected IP range, refusing.
```
* systemd-resolved is expecting the source ip to be in the range 127.0.0.0 to 127.255.255.255.255 (127.0.0.1/8).

Solutions:
* Tell zdns to use the local address 127.0.0.1: `zdns -local-addr 127.0.0.1 ......`
* Tell zdns to use another name server: `zdns -name-servers 8.8.8.8 .....`
* Tell systemd-resolved to allow that IP?
* Use some iptables rules to change the source IP?
* Update zdns to use the ip address of the interface it's using?
    * This may already be fixed in the current version of zdns, I haven't checked.

Good luck!

**Comment by phillip-stephens** [comment]
This was _partially_ addressed in #337, however we choose a "local address" based on the resolver. If for whatever reason you _mix_ resolvers in the input -
```
echo "cloudflare.com,127.0.0.53\nyahoo.com,1.1.1.1" | ./zdns A --verbosity 5
INFO[0000] No name servers specified. will use: 1.1.1.1:53
INFO[0000] using local address(es): [171.67.71.209]
DEBU[0000] DEPTH 01:    [****WIRE LOOKUP***  A   cloudflare.com   127.0.0.53:53]
DEBU[0000] DEPTH 01:    [****WIRE LOOKUP***  A   yahoo.com   1.1.1.1:53]
```

Then the local address will only work for the lookups it's in the same network as the resolver. So this is still a valid issue, IMO.

Also huge thanks @PenelopeFudd for the great write-up. I'll get this fixed.
